### ISSUE TITLE
Scrub never completes after rsyncing from .zfs/snapshot dir

### ISSUE THREAD

**Issue by brenc** [openzfs/zfs#16321]
### System information

Type | Version/Name
 --- | ---
Distribution Name	| FreeBSD
Distribution Version	| 14.1-RELEASE-p1
Kernel Version	| 14.1-RELEASE
Architecture	| x86_64
OpenZFS Version	| zfs-2.2.4-FreeBSD_g256659204, zfs-kmod-2.2.4-FreeBSD_g256659204

### Describe the problem you're observing
We have a remote backup process that uses rsync to back up files from a server using snapshots in the `.zfs/snapshot` directories. I noticed that after I started using this process subsequent scrubs (which run once per month on the 1st) never complete. I have been troubleshooting this for months and can confidently say that accessing files from these snapshot dirs is what triggers this bug. If I stop the remote backup process, reboot the server, and then run a scrub it completes normally in < 12 hours. As soon as I start the remote backup backup process, subsequent scrubs that run never complete even if left running for 5+ days.

I have tried pausing and resuming the scrub as well as stopping the scrub completely and restarting it. Nothing seems to correct this issue except a reboot. It always gets to around 70-80ish% done and hangs unless I reboot.

Some additional observations: scrub speed appears to be much reduced. This is what it looks like while stuck:
```
  pool: storage
 state: ONLINE
  scan: scrub in progress since Mon Jul  1 00:00:00 2024
        73.3T / 102T scanned at 595M/s, 73.3T / 102T issued at 595M/s
        0B repaired, 71.81% done, 14:05:39 to go
```
While a normal scrub will run at several T/s. There is definitely still high read IO on the drives:
```
                        capacity     operations     bandwidth    syncq_read    syncq_write   asyncq_read  asyncq_write   scrubq_read   trimq_write  rebuildq_write
pool                  alloc   free   read  write   read  write   pend  activ   pend  activ   pend  activ   pend  activ   pend  activ   pend  activ   pend  activ
--------------------  -----  -----  -----  -----  -----  -----  -----  -----  -----  -----  -----  -----  -----  -----  -----  -----  -----  -----  -----  -----
storage                102T   291T  4.87K      0  52.9M      0      0      0      0      0      0      0     97     48      0      0      0      0      0      0
  raidz2-0            25.6T  72.6T  1.19K      0  12.6M      0      0      0      0      0      0      0     18     12      0      0      0      0      0      0
    gpt/4BK7VW2H-zfs      -      -    192      0  1.88M      0      0      0      0      0      0      0      5      2      0      0      0      0      0      0
    gpt/4BK7WL5H-zfs      -      -    213      0  2.30M    807      0      0      0      0      0      0      2      2      0      0      0      0      0      0
    gpt/4YGV767H-zfs      -      -    180      0  1.72M    807      0      0      0      0      0      0      1      2      0      0      0      0      0      0
    gpt/4ZGRH7XV-zfs      -      -    203      0  1.99M    807      0      0      0      0      0      0      1      2      0      0      0      0      0      0
    gpt/4ZGRJENV-zfs      -      -    231      0  2.48M    807      0      0      0      0      0      0      1      2      0      0      0      0      0      0
    gpt/4ZGRJG9V-zfs      -      -    194      0  2.21M    807      0      0      0      0      0      0      2      2      0      0      0      0      0      0
  raidz2-1            25.6T  72.6T  1.27K      0  13.9M  3.94K      0      0      0      0      0      0     49     12      0      0      0      0      0      0
    gpt/4ZGRJGAV-zfs      -      -    202      0  2.34M    807      0      0      0      0      0      0      9      2      0      0      0      0      0      0
    gpt/4ZGRJJRV-zfs      -      -    214      0  2.26M    807      0      0      0      0      0      0      6      2      0      0      0      0      0      0
    gpt/4ZGRJJVV-zfs      -      -    202      0  1.94M    807      0      0      0      0      0      0      5      2      0      0      0      0      0      0
    gpt/4ZGRJKBV-zfs      -      -    198      0  1.88M  1.58K      0      0      0      0      0      0      2      2      0      0      0      0      0      0
    gpt/4ZGRJNTV-zfs      -      -    253      0  2.61M  1.58K      0      0      0      0      0      0      1      2      0      0      0      0      0      0
    gpt/4ZGRJSBV-zfs      -      -    226      0  2.87M  1.58K      0      0      0      0      0      0      9      2      0      0      0      0      0      0
  raidz2-2            25.6T  72.6T  1.10K      2  12.1M  9.47K      0      0      0      0      0      0     72     12      0      0      0      0      0      0
    gpt/4ZGRJTKV-zfs      -      -    211      0  2.18M  1.58K      0      0      0      0      0      0     12      2      0      0      0      0      0      0
    gpt/4ZGRJU3V-zfs      -      -    172      0  2.06M  1.58K      0      0      0      0      0      0     12      2      0      0      0      0      0      0
    gpt/4ZGRJU4V-zfs      -      -    189      0  2.17M  5.52K      0      0      0      0      0      0     10      2      0      0      0      0      0      0
    gpt/4ZGRJWJV-zfs      -      -    207      0  2.13M  1.58K      0      0      0      0      0      0     13      2      0      0      0      0      0      0
    gpt/4ZGRJXHV-zfs      -      -    161      0  1.76M  1.58K      0      0      0      0      0      0     13      2      0      0      0      0      0      0
    gpt/4ZGRJY3V-zfs      -      -    187      0  1.84M  1.58K      0      0      0      0      0      0      9      2      0      0      0      0      0      0
  raidz2-3            25.6T  72.6T  1.31K      4  14.3M  31.6K      0      0      0      0      0      0     15     12      0      0      0      0      0      0
    gpt/4ZGRJZ2V-zfs      -      -    205      0  2.01M  3.94K      0      0      0      0      0      0      2      2      0      0      0      0      0      0
    gpt/4ZGRJZBV-zfs      -      -    215      0  2.28M  3.94K      0      0      0      0      0      0      3      2      0      0      0      0      0      0
    gpt/4ZGRK4LV-zfs      -      -    235      0  2.46M  6.31K      0      0      0      0      0      0      5      2      0      0      0      0      0      0
    gpt/4ZGRK61V-zfs      -      -    218      0  2.37M  9.47K      0      0      0      0      0      0      3      2      0      0      0      0      0      0
    gpt/4ZGRKKYV-zfs      -      -    220      0  2.49M  3.94K      0      0      0      0      0      0      1      2      0      0      0      0      0      0
    gpt/4ZGRKLGV-zfs      -      -    242      0  2.67M  3.94K      0      0      0      0      0      0      1      2      0      0      0      0      0      0
--------------------  -----  -----  -----  -----  -----  -----  -----  -----  -----  -----  -----  -----  -----  -----  -----  -----  -----  -----  -----  -----
```
The scrub appears to be running even when no progress is made. % done never advances.

ZFS replication (this server pulls from the primary) essentially grinds to a halt as well.

Pool config:
```
# zpool get all storage
NAME     PROPERTY                       VALUE                          SOURCE
storage  size                           393T                           -
storage  capacity                       26%                            -
storage  altroot                        -                              default
storage  health                         ONLINE                         -
storage  guid                           15571001037900998829           -
storage  version                        -                              default
storage  bootfs                         -                              default
storage  delegation                     on                             default
storage  autoreplace                    off                            default
storage  cachefile                      -                              default
storage  failmode                       wait                           default
storage  listsnapshots                  off                            default
storage  autoexpand                     off                            default
storage  dedupratio                     1.00x                          -
storage  free                           291T                           -
storage  allocated                      102T                           -
storage  readonly                       off                            -
storage  ashift                         12                             local
storage  comment                        -                              default
storage  expandsize                     -                              -
storage  freeing                        0                              -
storage  fragmentation                  3%                             -
storage  leaked                         0                              -
storage  multihost                      off                            default
storage  checkpoint                     -                              -
storage  load_guid                      4612684389734077050            -
storage  autotrim                       off                            default
storage  compatibility                  off                            default
storage  bcloneused                     0                              -
storage  bclonesaved                    0                              -
storage  bcloneratio                    1.00x                          -
storage  feature@async_destroy          enabled                        local
storage  feature@empty_bpobj            active                         local
storage  feature@lz4_compress           active                         local
storage  feature@multi_vdev_crash_dump  enabled                        local
storage  feature@spacemap_histogram     active                         local
storage  feature@enabled_txg            active                         local
storage  feature@hole_birth             active                         local
storage  feature@extensible_dataset     active                         local
storage  feature@embedded_data          active                         local
storage  feature@bookmarks              enabled                        local
storage  feature@filesystem_limits      enabled                        local
storage  feature@large_blocks           active                         local
storage  feature@large_dnode            enabled                        local
storage  feature@sha512                 enabled                        local
storage  feature@skein                  enabled                        local
storage  feature@edonr                  enabled                        local
storage  feature@userobj_accounting     active                         local
storage  feature@encryption             active                         local
storage  feature@project_quota          active                         local
storage  feature@device_removal         enabled                        local
storage  feature@obsolete_counts        enabled                        local
storage  feature@zpool_checkpoint       enabled                        local
storage  feature@spacemap_v2            active                         local
storage  feature@allocation_classes     enabled                        local
storage  feature@resilver_defer         enabled                        local
storage  feature@bookmark_v2            enabled                        local
storage  feature@redaction_bookmarks    enabled                        local
storage  feature@redacted_datasets      enabled                        local
storage  feature@bookmark_written       enabled                        local
storage  feature@log_spacemap           active                         local
storage  feature@livelist               enabled                        local
storage  feature@device_rebuild         enabled                        local
storage  feature@zstd_compress          enabled                        local
storage  feature@draid                  enabled                        local
storage  feature@zilsaxattr             active                         local
storage  feature@head_errlog            active                         local
storage  feature@blake3                 enabled                        local
storage  feature@block_cloning          enabled                        local
storage  feature@vdev_zaps_v2           active                         local
```
I don't know if it matters but I'm using encryption. Here are the encryption root settings:
```
# zfs get all storage/e
NAME       PROPERTY              VALUE                           SOURCE
storage/e  type                  filesystem                      -
storage/e  creation              Wed Sep  6 15:10 2023           -
storage/e  used                  68.2T                           -
storage/e  available             193T                            -
storage/e  referenced            432K                            -
storage/e  compressratio         1.05x                           -
storage/e  mounted               yes                             -
storage/e  quota                 none                            default
storage/e  reservation           none                            default
storage/e  recordsize            1M                              inherited from storage
storage/e  mountpoint            /storage/e                      default
storage/e  sharenfs              off                             default
storage/e  checksum              on                              default
storage/e  compression           lz4                             inherited from storage
storage/e  atime                 off                             inherited from storage
storage/e  devices               on                              default
storage/e  exec                  off                             inherited from storage
storage/e  setuid                off                             inherited from storage
storage/e  readonly              off                             default
storage/e  jailed                off                             default
storage/e  snapdir               visible                         inherited from storage
storage/e  aclmode               discard                         default
storage/e  aclinherit            restricted                      default
storage/e  createtxg             62704                           -
storage/e  canmount              on                              default
storage/e  xattr                 on                              default
storage/e  copies                1                               default
storage/e  version               5                               -
storage/e  utf8only              off                             -
storage/e  normalization         none                            -
storage/e  casesensitivity       sensitive                       -
storage/e  vscan                 off                             default
storage/e  nbmand                off                             default
storage/e  sharesmb              off                             default
storage/e  refquota              none                            default
storage/e  refreservation        none                            default
storage/e  guid                  14413502527107633131            -
storage/e  primarycache          all                             default
storage/e  secondarycache        all                             default
storage/e  usedbysnapshots       0B                              -
storage/e  usedbydataset         432K                            -
storage/e  usedbychildren        68.2T                           -
storage/e  usedbyrefreservation  0B                              -
storage/e  logbias               latency                         default
storage/e  objsetid              282                             -
storage/e  dedup                 off                             default
storage/e  mlslabel              none                            default
storage/e  sync                  standard                        default
storage/e  dnodesize             legacy                          default
storage/e  refcompressratio      1.00x                           -
storage/e  written               432K                            -
storage/e  logicalused           70.7T                           -
storage/e  logicalreferenced     103K                            -
storage/e  volmode               default                         default
storage/e  filesystem_limit      none                            default
storage/e  snapshot_limit        none                            default
storage/e  filesystem_count      none                            default
storage/e  snapshot_count        none                            default
storage/e  snapdev               hidden                          default
storage/e  acltype               nfsv4                           default
storage/e  context               none                            default
storage/e  fscontext             none                            default
storage/e  defcontext            none                            default
storage/e  rootcontext           none                            default
storage/e  relatime              on                              default
storage/e  redundant_metadata    all                             default
storage/e  overlay               on                              default
storage/e  encryption            aes-256-gcm                     -
storage/e  keylocation           file:///etc/zfs_encryption_key  local
storage/e  keyformat             passphrase                      -
storage/e  pbkdf2iters           350000                          -
storage/e  encryptionroot        storage/e                       -
storage/e  keystatus             available                       -
storage/e  special_small_blocks  0                               default
storage/e  prefetch              all                             default
```

### Describe how to reproduce the problem
rsync files from `.zfs/snapshot` directories then run a scrub.

### Include any warning/errors/backtraces from the system logs
There are no relevant logs as far as I can tell.

I believe that's all the information I have available on this right now. Please let me know if you need anything else.

**Comment by awused** [comment]
I think I triggered this same bug but from cancelling a scrub instead, also on FreeBSD 14.1 with the exact same OpenZFS version.

I cancelled and restarted a scrub, then it seemed to get stuck at a certain point and the symptoms seem to match. The "issued" number and the completion percentage made no progress over about 12 hours before I gave up. The "scanned" number was going up but only by about 5-10MB/s for those same hours, and I was seeing 1-2k scrub read IOPS on the pool for the scrub but only 10-15MB/s of reads.

Restarting the scrub just caused it to get stuck at the same point again, and pausing/unpausing seemed to make a little progress but it just got stuck again slightly later. After restarting the machine the scrubbing is back to normal.

**Comment by brenc** [comment]
This is still happening for us. This month I paused the scrub and rebooted and even though the remote backup process immediately reconnected and started pulling data from the snapshot dirs, the scrub completed. I'm having to reboot once per month.

**Comment by exscape** [comment]
I seem to be having the same bug, but I've never figured out what triggers it. I have two similar systems with ZFS, and on one of them, the monthly scrub has gotten stuck *every single month* now for maybe 6-7 months in a row. On the other, it has only happened once -- today.  
Both run Debian bookworm with ZFS from backports.

Currently (on both systems):  
Kernel: 6.1.0-26-amd64 (Debian 6.1.112-1)  
ZFS: zfs-2.2.6-1~bpo12+2

Both are kept updated, so the this has happened on older kernels and ZFS versions as well.  
Both run ZFS native encryption (on the pool root). One is a 5-disk RAIDZ2 and the other (that hanged today) is a 2-disk mirror. The second system has a 1-disk zpool that is still scrubbing properly. 

It's possible that it's related to the snapshot dir for me as well, as I use it now and then... but I'm doubtful that I use it every single month, as I don't have any automated tools that use it. There hasn't been a monthly scrub on system 1 (RAIDZ2) that has finished by itself since possibly early this year.  
Pausing the scrub and rebooting always solves the issue, and as far as I can remember, I've never needed more than *one* reboot for a scrub to finish. It never seems to get stuck again, until next month.

The zpool status output sometimes progresses a miniscule amount, like 0.01% per 15+ minutes, but typically stands still. The (average) speed displayed always keeps going down towards 0 MB/s, and the remaining time always keeps going up, as no real progress is being made. The disks are still working quite hard, but I've no idea what they are actually doing.

This is what it looks like when it's working (iowait <2%, r/s >500 for most disks):

```
serenity@hyperion:~ $ iostat -mxy 30
Linux 6.1.0-23-amd64 (hyperion)         08/11/2024      _x86_64_        (8 CPU)

avg-cpu:  %user   %nice %system %iowait  %steal   %idle
           0.49    0.00    7.35    0.96    0.00   91.19

Device            r/s     rMB/s   rrqm/s  %rrqm r_await rareq-sz     w/s     wMB/s   wrqm/s  %wrqm w_await wareq-sz     d/s     dMB/s   drqm/s  %drqm d_await dareq-sz     f/s f_await  aqu-sz  %util
sda            922.60     65.89     1.27   0.14    0.86    73.13    9.43      0.08     0.00   0.00    3.24     8.51    0.00      0.00     0.00   0.00    0.00     0.00    0.33   88.10    0.85  34.25
sdb            904.43     65.88     1.30   0.14    0.89    74.59    9.27      0.08     0.00   0.00    3.48     8.83    0.00      0.00     0.00   0.00    0.00     0.00    0.33   93.00    0.87  35.00
sdc             68.33     65.98     0.00   0.00   41.71   988.76    9.57      0.08     0.00   0.00    1.77     8.56    0.00      0.00     0.00   0.00    0.00     0.00    0.33   46.20    2.88  96.53
sdd            882.97     65.90     1.50   0.17    1.01    76.43    9.43      0.08     0.00   0.00    3.08     8.71    0.00      0.00     0.00   0.00    0.00     0.00    0.33   83.30    0.94  37.19
sde            584.67     66.07     0.37   0.06    2.48   115.71    9.57      0.08     0.03   0.35    3.38     8.70    0.00      0.00     0.00   0.00    0.00     0.00    0.33   90.90    1.51  55.33

avg-cpu:  %user   %nice %system %iowait  %steal   %idle
           0.35    0.00    6.42    1.52    0.00   91.71

Device            r/s     rMB/s   rrqm/s  %rrqm r_await rareq-sz     w/s     wMB/s   wrqm/s  %wrqm w_await wareq-sz     d/s     dMB/s   drqm/s  %drqm d_await dareq-sz     f/s f_await  aqu-sz  %util
sda            745.07     63.06     1.03   0.14    1.46    86.67   10.17      0.09     0.00   0.00    3.79     9.13    0.00      0.00     0.00   0.00    0.00     0.00    0.40   86.92    1.16  43.68
sdb            754.57     63.09     0.80   0.11    1.43    85.62   10.37      0.09     0.00   0.00    2.86     9.02    0.00      0.00     0.00   0.00    0.00     0.00    0.40   66.67    1.13  42.37
sdc             84.13     63.31     0.00   0.00   33.47   770.56   10.63      0.09     0.00   0.00    1.84     8.59    0.00      0.00     0.00   0.00    0.00     0.00    0.40   43.75    2.85  95.44
sdd            664.97     63.07     1.00   0.15    2.01    97.12   10.27      0.09     0.00   0.00    3.29     8.90    0.00      0.00     0.00   0.00    0.00     0.00    0.40   76.08    1.40  50.20
sde            440.00     62.99     0.27   0.06    3.97   146.60   10.83      0.09     0.00   0.00    4.74     8.63    0.00      0.00     0.00   0.00    0.00     0.00    0.40   77.75    1.83  64.47

```

Below, it had stopped (iowait >10%, r/s <200 for all disks)

```
avg-cpu:  %user   %nice %system %iowait  %steal   %idle
           0.56    0.00    3.21   11.95    0.00   84.28

Device            r/s     rMB/s   rrqm/s  %rrqm r_await rareq-sz     w/s     wMB/s   wrqm/s  %wrqm w_await wareq-sz     d/s     dMB/s   drqm/s  %drqm d_await dareq-sz     f/s f_await  aqu-sz  %util
sda            161.87      5.57     0.00   0.00   17.10    35.23    4.57      0.05     0.00   0.00    4.91    10.39    0.00      0.00     0.00   0.00    0.00     0.00    0.20   89.83    2.81  93.91
sdb            159.80      5.70     0.00   0.00   17.75    36.53    4.67      0.05     0.00   0.00    4.03    10.06    0.00      0.00     0.00   0.00    0.00     0.00    0.20   77.83    2.87  95.64
sdc            150.60      5.50     0.00   0.00   15.60    37.40    5.13      0.05     0.03   0.65    1.86     9.35    0.00      0.00     0.00   0.00    0.00     0.00    0.20   41.00    2.37  78.81
sdd            162.23      5.72     0.00   0.00   17.85    36.10    5.00      0.05     0.00   0.00    4.00     9.52    0.00      0.00     0.00   0.00    0.00     0.00    0.20   92.67    2.93  97.68
sde            151.87      5.57     0.00   0.00   14.42    37.58    4.80      0.05     0.00   0.00    3.72    10.08    0.00      0.00     0.00   0.00    0.00     0.00    0.20   83.83    2.23  73.96

avg-cpu:  %user   %nice %system %iowait  %steal   %idle
           0.42    0.00    2.26   11.23    0.00   86.10

Device            r/s     rMB/s   rrqm/s  %rrqm r_await rareq-sz     w/s     wMB/s   wrqm/s  %wrqm w_await wareq-sz     d/s     dMB/s   drqm/s  %drqm d_await dareq-sz     f/s f_await  aqu-sz  %util
sda            166.73      5.85     0.00   0.00   17.29    35.92    3.50      0.03     0.00   0.00    3.99     9.07    0.00      0.00     0.00   0.00    0.00     0.00    0.13   84.00    2.91  96.89
sdb            165.30      6.00     0.00   0.00   17.31    37.17    3.53      0.03     0.07   1.85    3.27     8.83    0.00      0.00     0.00   0.00    0.00     0.00    0.13   79.75    2.88  96.15
sdc            163.43      5.78     0.00   0.00   15.55    36.24    3.70      0.03     0.00   0.00    1.54     8.29    0.00      0.00     0.00   0.00    0.00     0.00    0.13   33.50    2.55  84.71
sdd            166.33      5.96     0.00   0.00   17.54    36.69    3.80      0.03     0.00   0.00    3.67     8.39    0.00      0.00     0.00   0.00    0.00     0.00    0.13   92.25    2.94  98.04
sde            163.93      5.72     0.00   0.00   14.48    35.75    3.63      0.03     0.00   0.00    3.55     8.55    0.00      0.00     0.00   0.00    0.00     0.00    0.13   79.50    2.40  79.45
```

After a reboot, iostat returns to the pre-hang numbers where most disks have 500+ r/s and the scrub proceeds normally, finishing in maybe 12 hours. Without a reboot it won't ever finish.

This is what the second system looks like when the scrub is stuck. The mirror that is stuck is sdb+sdc; sda is a 1-disk zpool that doesn't need redundancy, and the scrub keeps working properly on that one. sdd is the boot SSD.

```
avg-cpu:  %user   %nice %system %iowait  %steal   %idle
           0.00    0.00    5.48   15.50    0.00   79.02

Device            r/s     rMB/s   rrqm/s  %rrqm r_await rareq-sz     w/s     wMB/s   wrqm/s  %wrqm w_await wareq-sz     d/s     dMB/s   drqm/s  %drqm d_await dareq-sz     f/s f_await  aqu-sz  %util
sda            173.00    171.57     0.00   0.00   17.26  1015.51    0.00      0.00     0.00   0.00    0.00     0.00    0.00      0.00     0.00   0.00    0.00     0.00    0.00    0.00    2.99  98.60
sdb            185.00     12.57     0.00   0.00   14.65    69.56   33.50      0.46     0.00   0.00    2.12    14.03    0.00      0.00     0.00   0.00    0.00     0.00    1.00   65.50    2.85  93.00
sdc            199.00     12.15     0.00   0.00   13.13    62.51   34.00      0.46     0.00   0.00    3.06    13.82    0.00      0.00     0.00   0.00    0.00     0.00    1.00   98.00    2.82  97.00
sdd              0.00      0.00     0.00   0.00    0.00     0.00    0.00      0.00     0.00   0.00    0.00     0.00    0.00      0.00     0.00   0.00    0.00     0.00    0.00    0.00    0.00   0.00
```

The scrubs haven't ever found any issues while this has happened, and AFAIK the pools keep working as usual during the stuck scrubs.

**Comment by awused** [comment]
This happened again on me, this time there was no obvious trigger. The uptime of the server is 38 days and it scrubs every monday. Usually it's almost done by now but, with my regular completion times hovering around 23-26 hours depending on activity. It was making almost no progress, had barely even scanned 1TB and hadn't even issued half a terabyte.

There was almost no other activity on the disks besides the scrub activity. It would issue ~15k reads, complete them very slowly, and then issue another 15k or so, but the actual disk activity was only 10-20MB/s.


```
 scan: scrub in progress since Mon Nov 18 03:00:00 2024
        1.17T / 33.8T scanned at 22.6M/s, 379G / 33.8T issued at 7.16M/s
        0B repaired, 1.10% done, no estimated completion time
```

**Comment by exscape** [comment]
I forgot to update for the last monthly scrub, but indeed the same issue happened on both computers again.  
My attempted workaround at pausing and resuming did *not* work, I had to reboot both computers for the scrubs to progress.

I'd really like some help in debugging this. I have no idea what data I can gather to help while the problem is occurring (except what's provided above, which presumably won't narrow down any exact cause).

**Comment by brenc** [comment]
I have to reboot our server every month. This month I forgot about it and the scrub went on for 12 days or so! Normally it finishes in 11-12 hours. I paused the scrub, rebooted, then restarted it and it finished in a matter of hours. I have yet to find a solution for this either.

**Comment by brenc** [comment]
Upgraded to 14.2-RELEASE. This bug is still present. I have to reboot every month.

**Comment by exscape** [comment]
Still keeps happening for me, also. But I might be having a different trigger for the same underlying issue, since I still doubt I've used the snapshot dir every single month, especially on the second system which mostly just sits there (100% of the storage is used for offsite backup, so I use the on-site ones if I need to restore something).

Still, I tried something new this month and checked some things with arcstat. I should add that I have no idea what I'm doing, but I haven't found any advice on how to get data on what all the extremely heavy disk accesses are actually doing, since they aren't scrubbing properly.

Before rebooting / when stuck:

```
hyperion ~ # arcstat -f time,hit%,dh%,ddh%,ddi%,dmh%,dmi%,mh%,mi%,ph%,pi%,pdh% 60
    time  hit%  dh%  ddh%  ddi%  dmh%  dmi%  mh%  mi%  ph%  pi%  pdh%
...
09:39:47    87  100   100     0   100     0   87    0   86    0     0
09:40:47    87  100   100     0   100     0   87    0   86    0     0
09:41:47    86  100   100     0   100     0   86    0   85    0     0
09:42:47    86  100   100     0   100     0   86    0   85    0     0
09:43:47    87  100   100     0   100     0   87    0   87    0     0
09:44:47    87  100   100     0   100     0   86    0   85    0     0
09:45:47    87   99   100     0   100     0   87    0   85    0     0

```

After rebooting (and waiting 15 minutes for the numbers to stabilize):

```
hyperion ~ # arcstat -f time,hit%,dh%,ddh%,ddi%,dmh%,dmi%,mh%,mi%,ph%,pi%,pdh% 60
    time  hit%  dh%  ddh%  ddi%  dmh%  dmi%  mh%  mi%  ph%  pi%  pdh%
10:23:58   100  100     0     0   100     0  100    0    0    0     0
10:24:58   100  100     0     0   100     0  100    0    0    0     0
10:25:58   100  100     0     0   100     0  100    0    0    0     0
10:26:58    64  100     0     0   100     0   64    0   45    0     0
10:27:58   100  100     0     0   100     0  100    0    0    0     0
10:28:58    68  100     0     0   100     0   68    0   46    0     0
10:29:58   100  100     0     0   100     0  100    0    0    0     0
10:30:58    99   99     0     0    99     0   99    0    0    0     0
10:31:58    93   99     0     0    99     0   93    0   66    0     0
10:32:58   100  100     0     0   100     0  100    0    0    0     0
10:33:58    81  100     0     0   100     0   81    0   68    0     0
10:34:58   100  100     0     0   100     0  100    0    0    0     0
```

So the biggest changes are ddh% (demand data hits percentage) which is basically always 100% when stuck and always 0% when scrubbing, and ph% (Prefetch hits percentage) which is 85-87% every time when stuck but usually 0 when scrubbing correctly. The same is true on the second system, about 92% when stuck and usually 0 when it works correctly.

Current versions:  
Debian 12.8  
Linux 6.1.0-28-amd64 (Debian package version 6.1.119-1)  
zfs-2.2.7-1~bpo12+1

**Comment by brenc** [comment]
@exscape thanks for posting this. I'll run the same commands during the next scrub. I had to reboot the box twice this month to get the scrub to complete.

**Comment by bubski** [comment]
I'm also experiencing scrubs slowing down to a crawl.

My system has 4 pools:
```
affected:
- sixpool: raidz2, 6 HDDs
unaffected:
- rpool mirror, 2 SSDs
- spool mirror, 2 SSDs
- ssdpool mirror, 2 SSDs
```

I first noticed a scrub on `sixpool` getting stuck around a month ago.
After a reboot it completed quickly.

I thought it was a one time thing, but yesterday it happened again.
It was progressing well until ~96%, and then pretty much stopped.
All 6 drives are being close to `100 %util` when checking with `iostat -x 5`.
`zpool iostat -v 2` also shows many reads, no drive stands out.

Pausing and resuming the scrub doesn't help.
`zpool scrub -p sixpool && zpool scrub sixpool`

Stopping and restarting the scrub doesn't help, and the same thing happens.
The scrub goes well until ~96% where it gets "stuck".
`zpool scrub -s sixpool && zpool scrub sixpool`

Not sure what else to check, but I'm pausing the scrub for now, and not rebooting the system just yet to try and gather more info.
I checked `arcstat` like @exscape above, and I'm seeing similar numbers.
I suspect I would see the same drop in `ddh%` and `ph%` after a reboot.
```
# arcstat -f time,hit%,dh%,ddh%,ddi%,dmh%,dmi%,mh%,mi%,ph%,pi%,pdh% 60
    time  hit%  dh%  ddh%  ddi%  dmh%  dmi%  mh%  mi%  ph%  pi%  pdh%
09:32:09     0    0     0     0     0     0    0    0    0    0     0
09:33:09    92  100   100     0   100     0   92    0   92    0     0
09:34:09    92  100   100     0   100     0   92    0   92    0     0
09:35:09    92  100   100     0   100     0   92    0   92    0     0
09:36:09    93  100   100     0   100     0   93    0   92    0     0
09:37:09    92  100   100     0   100     0   92    0   92    0     0
09:38:09    92   99   100     0   100     0   92    0   92    0     0
09:39:09    92  100   100     0   100     0   92    0   92    0     0
09:40:09    92  100   100     0   100     0   92    0   92    0     0
```

Some more info:
`sixpool` contains an encrypted datased (it's not the root dataset).
Same goes for `spool` and `ssdpool` though.

I updated from Proxmox 7 to 8 not long ago.
I'm pretty sure this was not happening before the update.

I have no idea what triggers this.
I see `.zfs/snapshot` mentioned in the thread, and while I touch those every now and then, I _don't think_ I have since the last reboot. But I'm not 100% certain.

Proxmox 8.3.3
Linux 6.8.12-8-pve
zfs-2.2.7-pve1

**Comment by davidcpgray** [comment]
I raised the issue #15885. There are some similarities but also some differences. I see the problem on a 4 disk raidz1 pool (HDD), but not on a 2 disk mirror (SSD). Both pools use native encryption. I hardly use snapshots at all on my system. 

The main difference I see here is that In my case the scrub doesn't even get past the initial 'scanning' phase to where it starts reporting data 'issued'. Lots of IOPS on the drives but no real 'throughput'. Zpool status never gets past 0% done.. 

My system is VERY lightly loaded and I am confident that it's not an IO contention problem or scrub prioritisation vs system work. There is mostly nothing else happening on the pool at the time of the scrub.  All the filesystems in the affected pool are encrypted. 

The workaround that I use is to export the pool then re-import the pool without re-loading the encryption keys. Then the pool scrubs successfully which takes around 15-16 hours for this system. No reboot required for this to work. Once the scrub is done I reload the keys and remount the filesystems. Appreciate this isn't necessarily possible in most use cases.

My system runs fully patched Debian bookworm with zfs-2.2.7-1~bpo12+1. I started seeing this behavior on Debian 11 and earlier versions of ZFS.

**Comment by awused** [comment]
I also use encryption for one dataset, but it's only a small portion of the pool (less than 10%). I've noticed sometimes a buggy scrub will eventually complete, but it'll take two or three times as long as a regular non-buggy scrub. Once the scrub gets past whatever is slowing it down the read speeds return to normal.

While the scrub is bugged, the pool is slow to respond to normal queries, as if it were very heavily loaded, and it can spend hours reading at <10MB/s, sometimes as low as a few hundred kilobytes per second, on a pool that usually gives >1GB/s reads for scrubs. There are far also more pending read operations when it's slow, and zfs waits for them to all complete before queuing another batch. I can see in iostat that there are no non-scrub operations for minutes at a time.

**Comment by exscape** [comment]
So apparently this can happen to resilvers as well... I tried to add a new disk yesterday, but when I booted up, another was missing. Long story short, the missing disk appears to have died, so instead of adding the new disk (RAIDZ expansion), I used `zpool replace` on it, to restore the RAIDZ2 to its original 5 disks.

When I went to bed yesterday, it looked like it would finish at 6 am or so. It's now 10 am and it's not progressing: 86.72% now, 86.72% 30 minutes ago. And iostat/arcstat show the same kind of numbers I've posted above when stuck during a scrub.

```
hyperion ~ # zpool status
  pool: laniakea
 state: DEGRADED
status: One or more devices is currently being resilvered.  The pool will
        continue to function, possibly in a degraded state.
action: Wait for the resilver to complete.
  scan: resilver in progress since Sun Apr 20 21:33:09 2025
        14.6T / 16.8T scanned at 336M/s, 14.6T / 16.8T issued at 335M/s
        2.92T resilvered, 86.72% done, 01:56:22 to go
config:

        NAME                                           STATE     READ WRITE CKSUM
        laniakea                                       DEGRADED     0     0     0
          raidz2-0                                     DEGRADED     0     0     0
            ata-WDC_WD40EFPX-68C6CN0_WD-...H2          ONLINE       0     0     0
            ata-WDC_WD40EFPX-68C6CN0_WD-...PS          ONLINE       0     0     0
            ata-WDC_WD40EFPX-68C6CN0_WD-...4Y          ONLINE       0     0     0
            replacing-3                                DEGRADED     0     0     0
              164...87                                 UNAVAIL      0     0     0  was /dev/disk/by-id/ata-WDC_WD40EZRX-00SPEB0_WD-WC...82-part1
              ata-HGST_HUS724040ALA640_...9S           ONLINE       0     0     0  (resilvering)
            ata-WDC_WD40EFRX-68N32N0_WD-WC...XL        ONLINE       0     0     0

errors: No known data errors
```

Standard "issued" speed is 550-600 MB/s. It's slowly going down as with the scrubs (looks like it's simply 14.6 T divided by the time taken, and the 14.6 isn't increasing).

```
hyperion # iostat -mxy 30
Linux 6.1.0-33-amd64 (hyperion)         04/21/2025      _x86_64_        (8 CPU)

avg-cpu:  %user   %nice %system %iowait  %steal   %idle
           0.62    0.00    3.99   23.90    0.00   71.50

Device            r/s     rMB/s   rrqm/s  %rrqm r_await rareq-sz     w/s     wMB/s   wrqm/s  %wrqm w_await wareq-sz     d/s     dMB/s   drqm/s  %drqm d_await dareq-sz     f/s f_await  aqu-sz  %util
sda            165.00      6.70     0.00   0.00   13.42    41.58    9.80      0.09     0.00   0.00    2.24     8.90    0.00      0.00     0.00   0.00    0.00     0.00    0.20   90.00    2.25  88.64
sdb            170.13      7.37     0.00   0.00   13.14    44.33    9.37      0.08     0.03   0.35    2.16     9.21    0.00      0.00     0.00   0.00    0.00     0.00    0.20   93.50    2.28  90.49
sdc              0.00      0.00     0.00   0.00    0.00     0.00   10.00      0.09     0.00   0.00    1.09     8.91    0.00      0.00     0.00   0.00    0.00     0.00    0.20   44.33    0.02   0.99
sdd            155.60      7.10     0.00   0.00   14.16    46.76    9.87      0.09     0.03   0.34    1.96     8.82    0.00      0.00     0.00   0.00    0.00     0.00    0.20   88.00    2.24  84.75
sde            160.13      6.73     0.00   0.00   12.17    43.01    9.83      0.09     0.00   0.00    2.28     8.91    0.00      0.00     0.00   0.00    0.00     0.00    0.20   89.50    1.99  79.67

avg-cpu:  %user   %nice %system %iowait  %steal   %idle
           0.40    0.00    3.03   20.58    0.00   75.98

Device            r/s     rMB/s   rrqm/s  %rrqm r_await rareq-sz     w/s     wMB/s   wrqm/s  %wrqm w_await wareq-sz     d/s     dMB/s   drqm/s  %drqm d_await dareq-sz     f/s f_await  aqu-sz  %util
sda            170.17      8.68     0.00   0.00   15.07    52.24   10.07      0.08     0.07   0.66    2.20     8.11    0.00      0.00     0.00   0.00    0.00     0.00    0.27   76.25    2.61  88.68
sdb            180.47      9.69     0.00   0.00   14.18    55.01    9.60      0.08     0.10   1.03    2.23     8.35    0.00      0.00     0.00   0.00    0.00     0.00    0.27   73.62    2.60  90.75
sdc              0.00      0.00     0.00   0.00    0.00     0.00    9.57      0.08     0.00   0.00    1.73     8.60    0.00      0.00     0.00   0.00    0.00     0.00    0.27   52.88    0.03   1.45
sdd            180.03      9.68     0.00   0.00   14.69    55.05   10.10      0.08     0.03   0.33    2.95     7.99    0.00      0.00     0.00   0.00    0.00     0.00    0.27  105.25    2.70  94.15
sde            172.20      8.35     0.00   0.00   11.68    49.68   10.07      0.08     0.00   0.00    2.64     8.04    0.00      0.00     0.00   0.00    0.00     0.00    0.27   92.12    2.06  73.55

hyperion # arcstat -f time,hit%,dh%,ddh%,ddi%,dmh%,dmi%,mh%,mi%,ph%,pi%,pdh% 60
    time  hit%  dh%  ddh%  ddi%  dmh%  dmi%  mh%  mi%  ph%  pi%  pdh%
10:03:23   0.0  0.0   0.0   0.0   0.0   0.0  0.0  0.0  0.0  0.0   0.0
10:04:23  87.4  100   100   0.0   100   0.0   87  0.0   70  0.0   0.0
10:05:23  73.4  100   100   0.0   100   0.0   72  0.0   68  0.0   0.0
10:06:23  72.8  100   100   0.0   100   0.1   72  0.0   70  0.0   0.0
10:07:23  73.6  100   100   0.0   100   0.1   72  0.0   70  0.0   0.0
10:08:23  73.5  100   100   0.0   100   0.1   72  0.0   71  0.0   0.0
```

Same as with the stuck scrubs, iowait is high (usually about 1% during a working scrub), r/s is <200, and in arcstat ddh% is 100% (0% when it works).
After reboot, everything progressed again. zpool status moved 3% in 20 minutes (vs 0.00% in 30 minutes prior to reboot), and iostat/arcstat show different numbers (iostat about 1%, r/s way above 200).

Sidenote, it looks like sdc above is now sdd (the disk being replaced in the RAIDZ2):

```
avg-cpu:  %user   %nice %system %iowait  %steal   %idle
           0.05    0.00    9.84    0.86    0.00   89.25

Device            r/s     rMB/s   rrqm/s  %rrqm r_await rareq-sz     w/s     wMB/s   wrqm/s  %wrqm w_await wareq-sz     d/s     dMB/s   drqm/s  %drqm d_await dareq-sz     f/s f_await  aqu-sz  %util
sda            503.20     84.60     1.07   0.21    3.77   172.16    7.73      0.07     0.10   1.28    2.91     9.78    0.00      0.00     0.00   0.00    0.00     0.00    0.33   57.40    1.94  66.16
sdb            319.87     84.69     0.23   0.07    7.19   271.12    8.00      0.08     0.03   0.41    3.32     9.63    0.00      0.00     0.00   0.00    0.00     0.00    0.33   64.40    2.35  79.97
sdc            487.47     84.64     0.87   0.18    3.99   177.81    8.30      0.08     0.03   0.40    2.22     9.30    0.00      0.00     0.00   0.00    0.00     0.00    0.33   51.60    1.98  67.40
sdd              0.10      0.00     0.00   0.00   60.67     9.33  265.73     83.91     0.77   0.29    6.58   323.35    0.00      0.00     0.00   0.00    0.00     0.00    0.33   56.10    1.77  87.81
sde            124.27     84.35     0.00   0.00   21.86   695.10    8.27      0.08     0.00   0.00    3.52     9.45    0.00      0.00     0.00   0.00    0.00     0.00    0.33   62.00    2.77  92.81


avg-cpu:  %user   %nice %system %iowait  %steal   %idle
           0.03    0.00   12.75    1.03    0.00   86.18

Device            r/s     rMB/s   rrqm/s  %rrqm r_await rareq-sz     w/s     wMB/s   wrqm/s  %wrqm w_await wareq-sz     d/s     dMB/s   drqm/s  %drqm d_await dareq-sz     f/s f_await  aqu-sz  %util
sda           1122.07     88.92     1.07   0.09    1.38    81.15    9.10      0.08     0.00   0.00    2.63     9.52    0.00      0.00     0.00   0.00    0.00     0.00    0.33   64.60    1.60  56.65
sdb           1032.70     88.77     0.77   0.07    1.54    88.02    9.73      0.09     0.00   0.00    3.03     9.07    0.00      0.00     0.00   0.00    0.00     0.00    0.33   72.70    1.64  57.75
sdc           1063.57     88.96     1.03   0.10    1.46    85.65    8.57      0.09     0.00   0.00    2.95    10.30    0.00      0.00     0.00   0.00    0.00     0.00    0.33   63.90    1.60  56.44
sdd              0.00      0.00     0.00   0.00    0.00     0.00  359.20     87.68     0.97   0.27    3.41   249.94    0.00      0.00     0.00   0.00    0.00     0.00    0.33   70.70    1.25  61.23
sde            232.37     89.71     0.10   0.04   10.36   395.36   10.27      0.09     0.00   0.00    3.08     8.68    0.00      0.00     0.00   0.00    0.00     0.00    0.33   80.60    2.47  83.11

hyperion ~ # arcstat -f time,hit%,dh%,ddh%,ddi%,dmh%,dmi%,mh%,mi%,ph%,pi%,pdh% 60
    time  hit%  dh%  ddh%  ddi%  dmh%  dmi%  mh%  mi%  ph%  pi%  pdh%
...
10:28:35   100  100   0.0   0.0   100   0.0  100  0.0  0.0  0.0   0.0
10:29:35   100  100   0.0   0.0   100   0.0  100  0.0  0.0  0.0   0.0
10:30:35   100  100   0.0   0.0   100   0.0  100  0.0  0.0  0.0   0.0
10:31:35   100  100   0.0   0.0   100   0.0  100  0.0  0.0  0.0   0.0
10:32:35   100  100   0.0   0.0   100   0.0  100  0.0  0.0  0.0   0.0
10:33:35   100  100   0.0   0.0   100   0.0  100  0.0  0.0  0.0   0.0
10:34:35   100  100   0.0   0.0   100   0.0  100  0.0  0.0  0.0   0.0
10:35:35   100  100   0.0   0.0   100   0.0  100  0.0  0.0  0.0   0.0
10:36:35   100  100   0.0   0.0   100   0.0  100  0.0  0.0  0.0   0.0
10:37:35   100  100   0.0   0.0   100   0.0  100  0.0  0.0  0.0   0.0
10:38:35   100  100   0.0   0.0   100   0.0  100  0.0  0.0  0.0   0.0
10:39:35  74.8   99   0.0   0.0  99.0   0.9   75  0.7   22  0.4   0.0
10:40:35  79.1   99   0.0   0.0  98.8   1.1   79  0.8   42  0.3   0.0
10:41:35  99.4   99   0.0   0.0  99.4   0.0   99  0.0  0.0  0.0   0.0
10:42:35  85.1   99   0.0   0.0  99.3   0.5   85  0.5   10  0.6   0.0
10:43:35   100  100   0.0   0.0   100   0.0  100  0.0  0.0  0.0   0.0
```

After a reboot, the resilver finished successfully (at about 12:05, so about the time it estimated 2 hours earlier).

**Comment by exscape** [comment]
Whatever the issue was for me, it seems resolved now. Either because of a ZFS upgrade to 2.3, or because I replaced the disk (see above) and then ran a RAIDZ expansion from 5-disk RAIDZ2 to 6-disk.  
I did that two months ago, and both times the monthly scrub finished successfully.

```
$ zfs --version
zfs-2.3.1-1~bpo12+1
zfs-kmod-2.3.1-1~bpo12+1
```

My second computer also had its scrub finish two months in a row, *however* that was on a new pool (2-disk mirror) that I created two months ago, and not the pool which previously got stuck.  
So something seems to happen with the pool itself that causes these issues. Creating a new pool solves it, and either a resilver+expansion or upgrade to 2.3.1 solved it for my other computer.

Sorry about the lack of detail, with so many possible explanations for the sudden fix.

**Comment by brenc** [comment]
This is still happening consistently for us. I upgrade and reboot every month otherwise the scrub will never finish.

```
[root@store01 ~]# freebsd-version -kru
14.2-RELEASE-p1
14.2-RELEASE-p1
14.2-RELEASE-p3
[root@store01 ~]# zfs --version
zfs-2.2.6-FreeBSD_g33174af15
zfs-kmod-2.2.6-FreeBSD_g33174af15
```

**Comment by brenc** [comment]
This happened on production this month. All I did was copy a few files from a `.zfs/snapshot` directory. Seems it doesn't take as much as I thought to trigger this bug. Had to reboot both servers (prod and backup as usual). Upgraded to FreeBSD 14.3-RELEASE which upgraded ZFS to 2.2.7 so we'll see if that makes a difference next month.

**Comment by brenc** [comment]
Got this on a resilver today. It got to about 50% done then froze. The disk that was supposed to be getting resilvered had zero activity but the rest were pretty active.

Then when I attempted a reboot, the system froze for over 20 minutes, so I had to hard reset the box. Super nerve-racking but thankfully it came back up and does appear to be continuing the resilver.

Plus after upgrading to 14.3-RELEASE, I get complete hangs every 4 days or so, so I had to configure watchdogd to automatically reboot the box (got a ticket open with FreeBSD for that). This might be a completely separate issue, but I'm kinda wondering if this is somehow related to this bug.

At this point, I'm going to disable these snapshot directories entirely. This feature is way too buggy for production use. Even copying 1-2 files can trigger this bug, and the only way to fix it is to reboot.

**Comment by bubski** [comment]
Still experiencing the issue on Proxmox 9 running ZFS 2.3.4.

```
$ zfs --version
zfs-2.3.4-pve1
zfs-kmod-2.3.4-pve1
```
```
$ pveversion
pve-manager/9.0.11/3bf5476b8a4699e2 (running kernel: 6.14.11-4-pve)
```

**Comment by brenc** [comment]
Still happening on FreeBSD 15.0-p1 and ZFS zfs-2.4.0-rc4-FreeBSD_g099f69ff5.
